upgrade/util/util.go

```diff
@@ -144,3 +148,770 @@ func CreateOrUpdateLonghornVersionSetting(namespace string, lhClient *lhclientse
+		bids = v.(map[string]*v1beta2.BackingImageDataSource)
+	} else {
+		bids = map[string]*v1beta2.BackingImageDataSource{}
+	}
+	bids[bid.Name] = obj
+	return obj, nil
+}
+
+func UpdateResources(namespace string, lhClient *lhclientset.Clientset, resourceMaps map[string]interface{}) error {
+	var err error
+
+	for resourceKind, resourceMap := range resourceMaps {
+		switch resourceKind {
+		case types.LonghornKindNode:
+			err = updateNodes(namespace, lhClient, resourceMap.(map[string]*longhorn.Node))
+		case types.LonghornKindVolume:
+			err = updateVolumes(namespace, lhClient, resourceMap.(map[string]*longhorn.Volume))
+		case types.LonghornKindEngine:
+			err = updateEngines(namespace, lhClient, resourceMap.(map[string]*longhorn.Engine))
+		case types.LonghornKindReplica:
+			err = updateReplicas(namespace, lhClient, resourceMap.(map[string]*longhorn.Replica))
+		case types.LonghornKindBackup:
+			err = updateBackups(namespace, lhClient, resourceMap.(map[string]*longhorn.Backup))
+		case types.LonghornKindEngineImage:
+			err = updateEngineImages(namespace, lhClient, resourceMap.(map[string]*longhorn.EngineImage))
+		case types.LonghornKindInstanceManager:
+			err = updateInstanceManagers(namespace, lhClient, resourceMap.(map[string]*longhorn.InstanceManager))
+		case types.LonghornKindShareManager:
+			err = updateShareManagers(namespace, lhClient, resourceMap.(map[string]*longhorn.ShareManager))
+		case types.LonghornKindBackingImage:
+			err = updateBackingImages(namespace, lhClient, resourceMap.(map[string]*longhorn.BackingImage))
+		case types.LonghornKindRecurringJob:
+			err = updateRecurringJobs(namespace, lhClient, resourceMap.(map[string]*longhorn.RecurringJob))
+		case types.LonghornKindSetting:
+			err = updateSettings(namespace, lhClient, resourceMap.(map[string]*longhorn.Setting))
+		default:
+			return fmt.Errorf("resource kind %v is not able to updated", resourceKind)
+		}
+
+		if err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func updateNodes(namespace string, lhClient *lhclientset.Clientset, nodes map[string]*longhorn.Node) error {
+	existingNodeList, err := lhClient.LonghornV1beta2().Nodes(namespace).List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+
+	for _, existingNode := range existingNodeList.Items {
+		node, ok := nodes[existingNode.Name]
+		if !ok {
+			continue
+		}
+
+		if !reflect.DeepEqual(existingNode.Status, node.Status) {
+			if _, err = lhClient.LonghornV1beta2().Nodes(namespace).UpdateStatus(context.TODO(), node, metav1.UpdateOptions{}); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
+				return err
+			}
+		}
+
+		if !reflect.DeepEqual(existingNode.Spec, node.Spec) ||
+			!reflect.DeepEqual(existingNode.ObjectMeta, node.ObjectMeta) {
+			if _, err = lhClient.LonghornV1beta2().Nodes(namespace).Update(context.TODO(), node, metav1.UpdateOptions{}); err != nil && !apierrors.IsConflict(errors.Cause(err)) {
+				return err
+			}
+		}
```

Comment:
Sometimes the upgrade involves label/annotation/ownership/... update. Checking spec & status may be not enough.
Similar below.

Reply:
Yeah. So, I also check `ObjectMeta` here. Anything I missed?

Reply:
OK, Sorry for missing that line.